# 데이터 명세서

## 1. 제공 파일

참가자에게 제공되는 파일은 아래와 같습니다.

| 구분 | 파일 | 행 수 | 내용 |
|---|---|---:|---|
| 학습 데이터 | `train/ldaps_train.csv` | 420,864 | 학습 기간의 LDAPS 기상 예보 데이터 |
| 학습 데이터 | `train/gfs_train.csv` | 236,736 | 학습 기간의 GFS 기상 예보 데이터 |
| 학습 데이터 | `train/train_labels.csv` | 26,304 | 학습 기간의 KPX 그룹별 실제 발전량 |
| 학습 데이터 | `train/scada_vestas_train.csv` | 157,819 | VESTAS 터빈 SCADA 실측 데이터 |
| 학습 데이터 | `train/scada_unison_train.csv` | 105,264 | UNISON 터빈 SCADA 실측 데이터 |
| 평가 데이터 | `test/ldaps_test.csv` | 140,160 | 평가 기간의 LDAPS 기상 예보 데이터 |
| 평가 데이터 | `test/gfs_test.csv` | 78,840 | 평가 기간의 GFS 기상 예보 데이터 |
| 제출 양식 | `sample_submission.csv` | 8,760 | 참가자가 예측값을 채워 제출할 CSV 양식 |
| 메타 정보 | `info.xlsx` | - | KPX 그룹, 터빈, 설비용량, 위치 정보 |
| 데이터 설명 | `data_description.md` | - | 제공 데이터 파일과 컬럼에 대한 설명 문서 |

모든 CSV 파일은 `UTF-8 with BOM`(`utf-8-sig`)으로 저장되어 있습니다.

시간은 모두 KST 기준이며, 형식은 `YYYY-MM-DD HH:MM:SS`입니다.

## 2. 예측 목표

참가자는 2025년 전체 기간에 대해 3개 KPX 그룹의 시간별 발전량을 예측합니다.

| 컬럼 | 의미 | 설비용량 |
|---|---|---:|
| `kpx_group_1` | KPX 그룹 1 발전량 | 21.6 MW |
| `kpx_group_2` | KPX 그룹 2 발전량 | 21.6 MW |
| `kpx_group_3` | KPX 그룹 3 발전량 | 21.0 MW |

예측해야 하는 발전량 단위는 `kWh`입니다.

설비용량 단위는 `MW`입니다. 평가에서 발전량과 설비용량을 함께 사용할 때는 시간 단위 발전량 기준에 맞춰 설비용량을 `kWh`로 환산해야 합니다.

1시간 기준 환산값은 아래와 같습니다.

| 컬럼 | 설비용량 | 1시간 기준 설비용량 |
|---|---:|---:|
| `kpx_group_1` | 21.6 MW | 21,600 kWh |
| `kpx_group_2` | 21.6 MW | 21,600 kWh |
| `kpx_group_3` | 21.0 MW | 21,000 kWh |

## 3. 기간 요약

| 파일 | 기간 |
|---|---|
| `ldaps_train.csv`, `gfs_train.csv` | 2022-01-01 01:00:00 ~ 2025-01-01 00:00:00 |
| `train_labels.csv` | 2022-01-01 01:00:00 ~ 2025-01-01 00:00:00 |
| `ldaps_test.csv`, `gfs_test.csv` | 2025-01-01 01:00:00 ~ 2026-01-01 00:00:00 |
| `sample_submission.csv` | 2025-01-01 01:00:00 ~ 2026-01-01 00:00:00 |

## 4. `info.xlsx`

KPX 그룹과 풍력발전기 메타 정보를 담은 파일입니다.

이 파일은 KPX 원천데이터 파일의 `info` 시트 내용을 제공합니다.

| 컬럼 | 의미 |
|---|---|
| `단계` | 풍력발전단지 단계 구분 |
| `명칭` | 발전단지 또는 발전기 위치 명칭 |
| `제작사` | 풍력터빈 제작사 |
| `모델명` | 풍력터빈 모델명 |
| `호기` | 터빈 번호 |
| `좌표(Google)` | Google 기준 터빈 위치 좌표 |
| `KPX그룹` | 해당 터빈이 속한 KPX 평가 그룹 |
| `Hub Height(m)` | 허브 높이 |
| `Rotor Diameter(m)` | 로터 직경 |
| `설비용량(MW)` | 개별 터빈 설비용량 |
| `그룹설비용량(MW)` | KPX 그룹별 총 설비용량 |

## 5. `train_labels.csv`

학습용 실제 발전량 데이터입니다.

| 컬럼 | 의미 |
|---|---|
| `kst_dtm` | 실제 발전량 집계 구간의 종료 시각입니다. |
| `kpx_group_1` | KPX 그룹 1 실제 발전량입니다. 단위는 `kWh`입니다. |
| `kpx_group_2` | KPX 그룹 2 실제 발전량입니다. 단위는 `kWh`입니다. |
| `kpx_group_3` | KPX 그룹 3 실제 발전량입니다. 단위는 `kWh`입니다. |

그룹별 Label 제공 기간은 아래와 같습니다.

| 컬럼 | 제공 기간 | 비고 |
|---|---|---|
| `kpx_group_1` | 2022년-2024년 | 2022년부터 값이 제공됩니다. |
| `kpx_group_2` | 2022년-2024년 | 2022년부터 값이 제공됩니다. |
| `kpx_group_3` | 2023년-2024년 | 2022년 구간은 빈칸입니다. |

`kst_dtm`은 1시간 단위입니다.

## 6. 기상 예보 데이터 공통 설명

대상 파일:

- `ldaps_train.csv`
- `ldaps_test.csv`
- `gfs_train.csv`
- `gfs_test.csv`

기상 예보 데이터는 매일 `09:00 KST`에 초기화된 예보자료를 사용합니다.

이 예보자료 중 다음날 `01:00:00`부터 그 다음날 `00:00:00`까지의 24시간 예보값만 추출해 날짜별로 이어 붙였습니다.

각 예보자료는 해당일 `13:00 KST`부터 사용 가능한 것으로 간주합니다.

| 컬럼 | 의미 |
|---|---|
| `forecast_kst_dtm` | 예보 대상 시각입니다. 발전량을 예측해야 하는 시각과 대응됩니다. |
| `data_available_kst_dtm` | 해당 예보 데이터가 사용 가능해진 시각입니다. |
| `grid_id` | 기상 격자 ID입니다. |
| `latitude` | 격자 위도입니다. |
| `longitude` | 격자 경도입니다. |

예:

| `forecast_kst_dtm` | `data_available_kst_dtm` |
|---|---|
| `2025-01-01 01:00:00` | `2024-12-31 13:00:00` |
| `2025-01-01 02:00:00` | `2024-12-31 13:00:00` |
| `2025-01-02 00:00:00` | `2024-12-31 13:00:00` |
| `2025-01-02 01:00:00` | `2025-01-01 13:00:00` |

즉 `01:00:00`부터 다음날 `00:00:00`까지의 24개 예보 대상 시각은 같은 `data_available_kst_dtm`을 갖습니다.

## 7. `ldaps_train.csv`, `ldaps_test.csv`

LDAPS 기상 예보 데이터입니다. 약 1.5 km 공간해상도의 16개 격자에 대해 제공됩니다.

`forecast_kst_dtm` 하나당 `grid_id` 16개 행이 존재합니다.

| 컬럼 | 의미 |
|---|---|
| `forecast_kst_dtm` | 예보 대상 시각 |
| `data_available_kst_dtm` | 예보 데이터 사용 가능 시각 |
| `grid_id` | LDAPS 격자 ID |
| `latitude` | 격자 위도 |
| `longitude` | 격자 경도 |
| `heightAboveGround_10_10u` | 지상 10 m U 성분 바람 |
| `heightAboveGround_10_10v` | 지상 10 m V 성분 바람 |
| `heightAboveGround_50_50MUmax` | 지상 50 m U 성분 바람 최댓값 |
| `heightAboveGround_50_50MUmin` | 지상 50 m U 성분 바람 최솟값 |
| `heightAboveGround_50_50MVmax` | 지상 50 m V 성분 바람 최댓값 |
| `heightAboveGround_50_50MVmin` | 지상 50 m V 성분 바람 최솟값 |
| `heightAboveGround_5_XBLWS` | 지상 5 m X 방향 경계층 바람 |
| `heightAboveGround_5_YBLWS` | 지상 5 m Y 방향 경계층 바람 |
| `heightAboveGround_2_t` | 지상 2 m 기온 |
| `heightAboveGround_2_dpt` | 지상 2 m 이슬점온도 |
| `heightAboveGround_2_r` | 지상 2 m 상대습도 |
| `heightAboveGround_2_q` | 지상 2 m 비습 |
| `surface_0_sp` | 지표면 기압 |
| `meanSea_0_prmsl` | 해면기압 |
| `etc_0_blh` | 경계층 높이 |
| `surface_0_NDNSW` | 지표면 순 하향 단파복사 |
| `surface_0_NDNLW` | 지표면 순 하향 장파복사 |
| `heightAboveGround_2_SWDIR` | 지상 2 m 직접 단파복사 |
| `heightAboveGround_2_SWDIF` | 지상 2 m 산란 단파복사 |
| `etc_0_hcc` | 상층운량 |
| `etc_0_mcc` | 중층운량 |
| `etc_0_lcc` | 하층운량 |
| `etc_0_VLCDC` | 매우 낮은 층 운량 |
| `surface_0_avg_lsprate` | 지표면 평균 대규모 강수율 |
| `surface_0_lssrate` | 지표면 대규모 강설률 |
| `surface_0_ncpcp` | 지표면 비대류성 강수량 |
| `surface_0_snol` | 지표면 적설 관련 변수 |
| `surface_0_SNOM` | 지표면 융설량 |
| `surface_0_lsm` | 육지/해양 마스크 |
| `surface_0_h` | 지표면 고도 |

## 8. `gfs_train.csv`, `gfs_test.csv`

GFS 기상 예보 데이터입니다. 약 0.25도 공간해상도의 9개 격자에 대해 제공됩니다.

`forecast_kst_dtm` 하나당 `grid_id` 9개 행이 존재합니다.

| 컬럼 | 의미 |
|---|---|
| `forecast_kst_dtm` | 예보 대상 시각 |
| `data_available_kst_dtm` | 예보 데이터 사용 가능 시각 |
| `grid_id` | GFS 격자 ID |
| `latitude` | 격자 위도 |
| `longitude` | 격자 경도 |
| `heightAboveGround_10_10u` | 지상 10 m U 성분 바람 |
| `heightAboveGround_10_10v` | 지상 10 m V 성분 바람 |
| `heightAboveGround_80_u` | 지상 80 m U 성분 바람 |
| `heightAboveGround_80_v` | 지상 80 m V 성분 바람 |
| `heightAboveGround_100_100u` | 지상 100 m U 성분 바람 |
| `heightAboveGround_100_100v` | 지상 100 m V 성분 바람 |
| `heightAboveGround_2_2t` | 지상 2 m 기온 |
| `heightAboveGround_2_2d` | 지상 2 m 이슬점온도 |
| `heightAboveGround_2_2r` | 지상 2 m 상대습도 |
| `heightAboveGround_2_2sh` | 지상 2 m 비습 |
| `planetaryBoundaryLayer_0_u` | 행성경계층 U 성분 바람 |
| `planetaryBoundaryLayer_0_v` | 행성경계층 V 성분 바람 |
| `planetaryBoundaryLayer_0_VRATE` | 행성경계층 수직 속도 |
| `surface_0_dswrf` | 지표면 하향 단파복사 플럭스 |
| `surface_0_dlwrf` | 지표면 하향 장파복사 플럭스 |
| `surface_0_prate` | 지표면 강수율 |
| `surface_0_tp` | 지표면 총강수량 |
| `surface_0_sp` | 지표면 기압 |
| `meanSea_0_prmsl` | 해면기압 |
| `surface_0_gust` | 지표면 돌풍 |
| `lowCloudLayer_0_lcc` | 하층운량 |
| `middleCloudLayer_0_mcc` | 중층운량 |
| `highCloudLayer_0_hcc` | 상층운량 |
| `atmosphere_0_tcc` | 전운량 |
| `isobaricInhPa_850_t` | 850 hPa 기온 |
| `isobaricInhPa_850_u` | 850 hPa U 성분 바람 |
| `isobaricInhPa_850_v` | 850 hPa V 성분 바람 |
| `isobaricInhPa_850_r` | 850 hPa 상대습도 |
| `isobaricInhPa_700_t` | 700 hPa 기온 |
| `isobaricInhPa_700_u` | 700 hPa U 성분 바람 |
| `isobaricInhPa_700_v` | 700 hPa V 성분 바람 |
| `isobaricInhPa_500_gh` | 500 hPa 지위고도 |
| `isobaricInhPa_500_t` | 500 hPa 기온 |
| `isobaricInhPa_500_u` | 500 hPa U 성분 바람 |
| `isobaricInhPa_500_v` | 500 hPa V 성분 바람 |

## 9. `scada_vestas_train.csv`

VESTAS 터빈 SCADA 학습 데이터입니다.

| 컬럼 | 의미 |
|---|---|
| `kst_dtm` | SCADA 계측 시각 |
| `vestas_wtg01_power_kw10m` ~ `vestas_wtg12_power_kw10m` | VESTAS 1-12호기 10분 단위 power 값입니다. 원천 컬럼명 기준 단위는 `kW10m`입니다. |
| `vestas_wtg01_ws` ~ `vestas_wtg12_ws` | VESTAS 1-12호기 풍속입니다. |
| `vestas_wtg01_wd` ~ `vestas_wtg12_wd` | VESTAS 1-12호기 풍향입니다. |

## 10. `scada_unison_train.csv`

UNISON 터빈 SCADA 학습 데이터입니다.

| 컬럼 | 의미 |
|---|---|
| `kst_dtm` | SCADA 계측 시각 |
| `unison_wtg01_power_kw10m` ~ `unison_wtg05_power_kw10m` | UNISON 1-5호기 10분 단위 power 값입니다. 원천 컬럼명 기준 단위는 `kW10m`입니다. |
| `unison_wtg01_ws` ~ `unison_wtg05_ws` | UNISON 1-5호기 풍속입니다. |
| `unison_wtg01_wd` ~ `unison_wtg05_wd` | UNISON 1-5호기 풍향입니다. |

## 11. `sample_submission.csv`

제출 양식 파일입니다.

| 컬럼 | 의미 |
|---|---|
| `forecast_id` | 제출 행을 구분하는 고유 ID입니다. 평가 시 매칭 키로 사용됩니다. |
| `forecast_kst_dtm` | 예측 대상 시각입니다. |
| `kpx_group_1` | KPX 그룹 1 발전량 예측값을 입력합니다. 단위는 `kWh`입니다. |
| `kpx_group_2` | KPX 그룹 2 발전량 예측값을 입력합니다. 단위는 `kWh`입니다. |
| `kpx_group_3` | KPX 그룹 3 발전량 예측값을 입력합니다. 단위는 `kWh`입니다. |

참가자는 `forecast_id`와 `forecast_kst_dtm`을 변경하지 말고, `kpx_group_1`, `kpx_group_2`, `kpx_group_3` 값만 채워 제출해야 합니다.

## 12. 주의사항

평가 기간의 실제 발전량, SCADA, 비공개 운영 데이터, 사후 보정자료, 재분석자료 등은 예측에 사용할 수 없습니다.

Excel에서 CSV 파일을 열고 다시 저장하면 시간 문자열 형식이 바뀔 수 있습니다. 제출 파일은 코드로 생성하는 것을 권장합니다.


In [1]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
DATA_DIR = Path("../datasets")
TRAIN_DIR = DATA_DIR / "train"
TEST_DIR = DATA_DIR / "test"

TARGET_COLS = ["kpx_group_1", "kpx_group_2", "kpx_group_3"]
CAPACITY_KWH = {
    "kpx_group_1": 21600,
    "kpx_group_2": 21600,
    "kpx_group_3": 21000,
}

train_labels = pd.read_csv(TRAIN_DIR / "train_labels.csv", encoding="utf-8-sig")
sample_submission = pd.read_csv(DATA_DIR / "sample_submission.csv", encoding="utf-8-sig")

ldaps_train = pd.read_csv(TRAIN_DIR / "ldaps_train.csv", encoding="utf-8-sig")
gfs_train = pd.read_csv(TRAIN_DIR / "gfs_train.csv", encoding="utf-8-sig")
ldaps_test = pd.read_csv(TEST_DIR / "ldaps_test.csv", encoding="utf-8-sig")
gfs_test = pd.read_csv(TEST_DIR / "gfs_test.csv", encoding="utf-8-sig")

scada_vestas_train = pd.read_csv(TRAIN_DIR / "scada_vestas_train.csv", encoding="utf-8-sig")
scada_unison_train = pd.read_csv(TRAIN_DIR / "scada_unison_train.csv", encoding="utf-8-sig")

# info.xlsx: 상단 3행은 제목/공백이라 스킵하고, 병합 셀로 비어있는 KPX그룹/그룹설비용량(MW)은 ffill로 채움
info_df = pd.read_excel(DATA_DIR / "info.xlsx", skiprows=3)
info_df = info_df.dropna(axis=1, how="all").ffill()

train_labels["kst_dtm"] = pd.to_datetime(train_labels["kst_dtm"])
sample_submission["forecast_kst_dtm"] = pd.to_datetime(sample_submission["forecast_kst_dtm"])

for _df in (ldaps_train, gfs_train, ldaps_test, gfs_test):
    _df["forecast_kst_dtm"] = pd.to_datetime(_df["forecast_kst_dtm"])
    _df["data_available_kst_dtm"] = pd.to_datetime(_df["data_available_kst_dtm"])

for _df in (scada_vestas_train, scada_unison_train):
    _df["kst_dtm"] = pd.to_datetime(_df["kst_dtm"])

print("train_labels:", train_labels.shape)
print("sample_submission:", sample_submission.shape)
print("ldaps_train:", ldaps_train.shape, "gfs_train:", gfs_train.shape)
print("ldaps_test:", ldaps_test.shape, "gfs_test:", gfs_test.shape)
print("scada_vestas_train:", scada_vestas_train.shape)
print("scada_unison_train:", scada_unison_train.shape)
print("info_df:", info_df.shape)

train_labels: (26304, 4)
sample_submission: (8760, 5)
ldaps_train: (420864, 35) gfs_train: (236736, 40)
ldaps_test: (140160, 35) gfs_test: (78840, 40)
scada_vestas_train: (157819, 37)
scada_unison_train: (105264, 16)
info_df: (17, 11)


In [3]:
train_label_df = pd.DataFrame(train_labels)
train_label_df.head()

,kst_dtm,kpx_group_1,kpx_group_2,kpx_group_3
0,2022-01-01 01:00:00,12004.421,9719.242,NaN
1,2022-01-01 02:00:00,12901.137,10297.768,NaN
2,2022-01-01 03:00:00,12091.200,10731.663,NaN
3,2022-01-01 04:00:00,17167.768,13841.242,NaN
4,2022-01-01 05:00:00,19134.758,14014.800,NaN


In [4]:
train_label_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 26304 entries, 0 to 26303
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   kst_dtm      26304 non-null  datetime64[us]
 1   kpx_group_1  26200 non-null  float64       
 2   kpx_group_2  26201 non-null  float64       
 3   kpx_group_3  17538 non-null  float64       
dtypes: datetime64[us](1), float64(3)
memory usage: 822.1 KB


## train_label 데이터 이슈파악

In [5]:
print('데이터 크기:', train_label_df.shape)
display(train_label_df.head())
print('[자료형 및 결측치 요약]')
display(train_label_df.info())
print('[기초 통계량]')
display(train_label_df.describe())
print('[결측치 개수]')
display(train_label_df.isnull().sum())
print('[결측치 비율]')
display((train_label_df.isnull().mean() * 100).round(2))
print('[중복 행 개수]', train_label_df.duplicated().sum())
print('[컬럼별 고유값 개수]')
display(train_label_df.nunique())

numeric_cols = train_label_df.select_dtypes(include='number').columns
categorical_cols = train_label_df.select_dtypes(exclude='number').columns

print('숫자형 컬럼:', list(numeric_cols))
print('범주형 컬럼:', list(categorical_cols))
print('[숫자형 컬럼 스케일 확인]')
display(train_label_df[numeric_cols].agg(['min', 'max', 'mean', 'std']).T)

Q1 = train_label_df[numeric_cols].quantile(0.25)
Q3 = train_label_df[numeric_cols].quantile(0.75)
IQR = Q3 - Q1
outlier_count = ((train_label_df[numeric_cols] < (Q1 - 1.5 * IQR)) | (train_label_df[numeric_cols] > (Q3 + 1.5 * IQR))).sum()
print('[IQR 기준 이상치 후보 개수]')
display(outlier_count)

데이터 크기: (26304, 4)


,kst_dtm,kpx_group_1,kpx_group_2,kpx_group_3
0,2022-01-01 01:00:00,12004.421,9719.242,NaN
1,2022-01-01 02:00:00,12901.137,10297.768,NaN
2,2022-01-01 03:00:00,12091.200,10731.663,NaN
3,2022-01-01 04:00:00,17167.768,13841.242,NaN
4,2022-01-01 05:00:00,19134.758,14014.800,NaN


[자료형 및 결측치 요약]
<class 'pandas.DataFrame'>
RangeIndex: 26304 entries, 0 to 26303
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   kst_dtm      26304 non-null  datetime64[us]
 1   kpx_group_1  26200 non-null  float64       
 2   kpx_group_2  26201 non-null  float64       
 3   kpx_group_3  17538 non-null  float64       
dtypes: datetime64[us](1), float64(3)
memory usage: 822.1 KB


None

[기초 통계량]


,kst_dtm,kpx_group_1,kpx_group_2,kpx_group_3
count,26304,26200.000000,26201.000000,17538.000000
mean,2023-07-03 00:30:00,6621.981125,7076.842859,5563.819649
min,2022-01-01 01:00:00,0.000000,0.000000,0.000000
25%,2022-10-02 00:45:00,549.600000,549.600000,159.095000
50%,2023-07-03 00:30:00,4252.168000,4382.337000,2719.074000
75%,2024-04-02 00:15:00,12206.905000,13508.589000,9979.579000
max,2025-01-01 00:00:00,21275.305000,21362.084000,21130.674000
std,NaN,6582.443128,7001.146068,6294.582901


[결측치 개수]


kst_dtm           0
kpx_group_1     104
kpx_group_2     103
kpx_group_3    8766
dtype: int64

[결측치 비율]


kst_dtm         0.00
kpx_group_1     0.40
kpx_group_2     0.39
kpx_group_3    33.33
dtype: float64

[중복 행 개수] 0
[컬럼별 고유값 개수]


kst_dtm        26304
kpx_group_1     1472
kpx_group_2     1478
kpx_group_3     1459
dtype: int64

숫자형 컬럼: ['kpx_group_1', 'kpx_group_2', 'kpx_group_3']
범주형 컬럼: ['kst_dtm']
[숫자형 컬럼 스케일 확인]


,min,max,mean,std
kpx_group_1,0.0,21275.305,6621.981125,6582.443128
kpx_group_2,0.0,21362.084,7076.842859,7001.146068
kpx_group_3,0.0,21130.674,5563.819649,6294.582901


[IQR 기준 이상치 후보 개수]


kpx_group_1    0
kpx_group_2    0
kpx_group_3    0
dtype: int64

## 공통 이슈파악 함수 정의

이후 각 데이터셋(LDAPS/GFS 예보, SCADA)에 반복 적용할 EDA 유틸리티 함수입니다.

In [6]:
def summarize_basic(df: pd.DataFrame, name: str) -> None:
    print(f"===== {name} 기본 EDA =====")
    print("데이터 크기:", df.shape)
    display(df.head())
    print("[dtypes]")
    display(df.dtypes)

    print("[결측치 개수 / 비율(%) - 결측 존재 컬럼만]")
    na_summary = pd.concat(
        [df.isnull().sum(), (df.isnull().mean() * 100).round(2)],
        axis=1, keys=["count", "pct"],
    )
    display(na_summary[na_summary["count"] > 0])
    print("[중복 행 개수]", df.duplicated().sum())

    numeric_cols = df.select_dtypes(include="number").columns
    if len(numeric_cols) == 0:
        return

    print("[숫자형 컬럼 스케일 확인]")
    display(df[numeric_cols].describe().T)

    Q1 = df[numeric_cols].quantile(0.25)
    Q3 = df[numeric_cols].quantile(0.75)
    IQR = Q3 - Q1
    outlier_count = (
        (df[numeric_cols] < (Q1 - 1.5 * IQR)) | (df[numeric_cols] > (Q3 + 1.5 * IQR))
    ).sum()
    print("[IQR 기준 이상치 후보 개수 (상위 10개 컬럼)]")
    display(outlier_count.sort_values(ascending=False).head(10))


def check_forecast_grid(df: pd.DataFrame, name: str, expected_grid_count: int) -> None:
    print(f"===== {name} 격자/시각 정합성 =====")
    grid_ids = sorted(df["grid_id"].unique())
    print(f"grid_id 개수: {len(grid_ids)} (기대값 {expected_grid_count})")

    per_time_grid_count = df.groupby("forecast_kst_dtm")["grid_id"].nunique()
    bad_times = per_time_grid_count[per_time_grid_count != expected_grid_count]
    print(f"grid_id 개수가 기대값과 다른 forecast_kst_dtm 시각 수: {len(bad_times)}")
    if len(bad_times):
        display(bad_times.head())

    dup_key = df.duplicated(subset=["forecast_kst_dtm", "grid_id"]).sum()
    print("(forecast_kst_dtm, grid_id) 조합 중복 개수:", dup_key)

    latlon_nunique = df.groupby("grid_id")[["latitude", "longitude"]].nunique()
    moving_grids = latlon_nunique[(latlon_nunique["latitude"] > 1) | (latlon_nunique["longitude"] > 1)]
    print("격자별 좌표가 고정되지 않은 grid_id 개수:", len(moving_grids))

    full_hours = pd.date_range(df["forecast_kst_dtm"].min(), df["forecast_kst_dtm"].max(), freq="h")
    missing_hours = full_hours.difference(df["forecast_kst_dtm"].unique())
    print(f"기간 내 누락된 forecast_kst_dtm 시각 수: {len(missing_hours)} / 전체 {len(full_hours)}")

    block_sizes = df.groupby(["grid_id", "data_available_kst_dtm"])["forecast_kst_dtm"].count()
    print("[data_available_kst_dtm 블록 크기 분포 (정상이면 24시간 단위)]")
    display(block_sizes.value_counts().sort_index())


def check_scada(df: pd.DataFrame, name: str, prefix: str, turbine_ids: list, freq_minutes: int = 10) -> None:
    print(f"===== {name} SCADA 이슈파악 =====")
    dt_col = "kst_dtm"
    print("타임스탬프 중복 개수:", df[dt_col].duplicated().sum())

    full_range = pd.date_range(df[dt_col].min(), df[dt_col].max(), freq=f"{freq_minutes}min")
    missing_ts = full_range.difference(df[dt_col])
    print(f"기간 내 누락된 타임스탬프 수: {len(missing_ts)} / 전체 {len(full_range)}")

    rows = []
    for tid in turbine_ids:
        power_col = f"{prefix}_{tid}_power_kw10m"
        ws_col = f"{prefix}_{tid}_ws"
        wd_col = f"{prefix}_{tid}_wd"
        power = df[power_col]

        same_as_prev = power.notna() & (power == power.shift())
        max_run = same_as_prev.groupby((~same_as_prev).cumsum()).transform("size").where(same_as_prev).max()

        rows.append({
            "turbine": tid,
            "결측치": power.isnull().sum(),
            "음수 power": int((power < 0).sum()),
            "동일값 최대 연속 길이": 0 if pd.isna(max_run) else int(max_run) + 1,
            "풍속0인데 power>0 건수": int(((df[ws_col] == 0) & (power > 0)).sum()),
            "풍향 범위": f"{df[wd_col].min():.1f} ~ {df[wd_col].max():.1f}",
        })
    display(pd.DataFrame(rows).set_index("turbine"))

## ldaps_train 데이터 이슈파악

In [7]:
summarize_basic(ldaps_train, "ldaps_train")
check_forecast_grid(ldaps_train, "ldaps_train", expected_grid_count=16)

===== ldaps_train 기본 EDA =====
데이터 크기: (420864, 35)


,forecast_kst_dtm,data_available_kst_dtm,grid_id,latitude,longitude,heightAboveGround_10_10u,heightAboveGround_10_10v,heightAboveGround_50_50MUmax,heightAboveGround_50_50MUmin,heightAboveGround_50_50MVmax,...,etc_0_mcc,etc_0_lcc,etc_0_VLCDC,surface_0_avg_lsprate,surface_0_lssrate,surface_0_ncpcp,surface_0_snol,surface_0_SNOM,surface_0_lsm,surface_0_h
0,2022-01-01 01:00:00,2021-12-31 13:00:00,1,37.3032,128.9443,5.411171,1.591679,8.162608,7.618145,2.164301,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,992.6250
1,2022-01-01 01:00:00,2021-12-31 13:00:00,2,37.3027,128.9617,5.593300,0.727909,8.908702,8.205059,1.596918,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,936.5625
2,2022-01-01 01:00:00,2021-12-31 13:00:00,3,37.3022,128.9790,4.383828,-0.924435,7.793468,7.167461,0.100824,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,868.8125
3,2022-01-01 01:00:00,2021-12-31 13:00:00,4,37.2899,128.9263,3.022011,0.641972,5.043956,4.645001,0.879145,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,926.5000
4,2022-01-01 01:00:00,2021-12-31 13:00:00,5,37.2894,128.9437,4.837441,0.774784,7.277843,6.799786,1.086176,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,997.6250


[dtypes]


forecast_kst_dtm                datetime64[us]
data_available_kst_dtm          datetime64[us]
grid_id                                  int64
latitude                               float64
longitude                              float64
heightAboveGround_10_10u               float64
heightAboveGround_10_10v               float64
heightAboveGround_50_50MUmax           float64
heightAboveGround_50_50MUmin           float64
heightAboveGround_50_50MVmax           float64
heightAboveGround_50_50MVmin           float64
heightAboveGround_5_XBLWS              float64
heightAboveGround_5_YBLWS              float64
heightAboveGround_2_t                  float64
heightAboveGround_2_dpt                float64
heightAboveGround_2_r                  float64
heightAboveGround_2_q                  float64
surface_0_sp                           float64
meanSea_0_prmsl                        float64
etc_0_blh                              float64
surface_0_NDNSW                        float64
surface_0_NDN

[결측치 개수 / 비율(%) - 결측 존재 컬럼만]


,count,pct


[중복 행 개수] 0
[숫자형 컬럼 스케일 확인]


,count,mean,std,min,25%,50%,75%,max
grid_id,420864.0,8.500000,4.609778,1.000000e+00,4.750000,8.500000,12.250000,16.000000
latitude,420864.0,37.281931,0.013849,3.726070e+01,37.274375,37.281900,37.289525,37.303200
longitude,420864.0,128.960719,0.021274,1.289257e+02,128.943525,128.960700,128.977950,128.995800
heightAboveGround_10_10u,420864.0,2.738978,4.332892,-1.182553e+01,-1.140903,3.284855,5.853243,18.085985
heightAboveGround_10_10v,420864.0,0.425384,1.938003,-1.734487e+01,-0.623062,0.350500,1.452382,12.711470
heightAboveGround_50_50MUmax,420864.0,4.696355,6.560151,-1.720542e+01,-0.714551,5.166576,9.427945,29.133512
heightAboveGround_50_50MUmin,420864.0,3.605434,6.554589,-1.955014e+01,-2.081963,4.165522,8.414210,27.475353
heightAboveGround_50_50MVmax,420864.0,1.145152,3.049756,-2.469426e+01,-0.447014,0.929294,2.536415,20.772963
heightAboveGround_50_50MVmin,420864.0,0.156378,2.954496,-2.777414e+01,-1.286951,0.140192,1.602886,18.794529
heightAboveGround_5_XBLWS,420864.0,0.024902,0.080477,-5.353961e-01,-0.006596,0.007798,0.042192,0.957414


[IQR 기준 이상치 후보 개수 (상위 10개 컬럼)]


etc_0_VLCDC                  80856
etc_0_mcc                    67884
heightAboveGround_2_SWDIR    62978
heightAboveGround_5_XBLWS    54190
heightAboveGround_5_YBLWS    53746
surface_0_avg_lsprate        51044
surface_0_SNOM               47131
surface_0_ncpcp              43413
surface_0_NDNSW              31211
etc_0_blh                    29032
dtype: int64

===== ldaps_train 격자/시각 정합성 =====
grid_id 개수: 16 (기대값 16)
grid_id 개수가 기대값과 다른 forecast_kst_dtm 시각 수: 0
(forecast_kst_dtm, grid_id) 조합 중복 개수: 0
격자별 좌표가 고정되지 않은 grid_id 개수: 0
기간 내 누락된 forecast_kst_dtm 시각 수: 0 / 전체 26304
[data_available_kst_dtm 블록 크기 분포 (정상이면 24시간 단위)]


forecast_kst_dtm
24    17536
Name: count, dtype: int64

## gfs_train 데이터 이슈파악

In [8]:
summarize_basic(gfs_train, "gfs_train")
check_forecast_grid(gfs_train, "gfs_train", expected_grid_count=9)

===== gfs_train 기본 EDA =====
데이터 크기: (236736, 40)


,forecast_kst_dtm,data_available_kst_dtm,grid_id,latitude,longitude,heightAboveGround_10_10u,heightAboveGround_10_10v,heightAboveGround_80_u,heightAboveGround_80_v,heightAboveGround_100_100u,...,isobaricInhPa_850_u,isobaricInhPa_850_v,isobaricInhPa_850_r,isobaricInhPa_700_t,isobaricInhPa_700_u,isobaricInhPa_700_v,isobaricInhPa_500_gh,isobaricInhPa_500_t,isobaricInhPa_500_u,isobaricInhPa_500_v
0,2022-01-01 01:00:00,2021-12-31 13:00:00,1,37.50,128.75,1.817466,1.040068,2.437944,1.258201,2.539397,...,5.928218,-3.632051,8.5,260.32092,11.584992,-18.589058,5474.115,243.86014,22.805730,-15.617361
1,2022-01-01 01:00:00,2021-12-31 13:00:00,2,37.50,129.00,2.947466,1.340068,3.747944,1.728201,3.839397,...,4.158218,-2.582051,8.3,259.80090,11.744992,-18.769058,5470.083,244.00014,23.905731,-16.017360
2,2022-01-01 01:00:00,2021-12-31 13:00:00,3,37.50,129.25,2.817466,0.800068,2.577944,0.808201,2.509397,...,4.228218,-2.532051,9.5,259.58093,12.354993,-18.999058,5465.347,244.09013,24.905731,-16.317362
3,2022-01-01 01:00:00,2021-12-31 13:00:00,4,37.25,128.75,1.227466,0.190068,2.037944,0.038201,2.159397,...,5.308218,-4.312051,8.2,260.77090,11.854993,-18.309057,5480.051,244.00014,23.605732,-14.717361
4,2022-01-01 01:00:00,2021-12-31 13:00:00,5,37.25,129.00,2.457466,0.540068,3.107944,0.568201,3.129397,...,3.528218,-3.132051,7.8,260.33093,11.714993,-18.689058,5476.259,244.14014,24.205730,-15.317362


[dtypes]


forecast_kst_dtm                  datetime64[us]
data_available_kst_dtm            datetime64[us]
grid_id                                    int64
latitude                                 float64
longitude                                float64
heightAboveGround_10_10u                 float64
heightAboveGround_10_10v                 float64
heightAboveGround_80_u                   float64
heightAboveGround_80_v                   float64
heightAboveGround_100_100u               float64
heightAboveGround_100_100v               float64
heightAboveGround_2_2t                   float64
heightAboveGround_2_2d                   float64
heightAboveGround_2_2r                   float64
heightAboveGround_2_2sh                  float64
planetaryBoundaryLayer_0_u               float64
planetaryBoundaryLayer_0_v               float64
planetaryBoundaryLayer_0_VRATE           float64
surface_0_dswrf                          float64
surface_0_dlwrf                          float64
surface_0_prate     

[결측치 개수 / 비율(%) - 결측 존재 컬럼만]


,count,pct


[중복 행 개수] 0
[숫자형 컬럼 스케일 확인]


,count,mean,std,min,25%,50%,75%,max
grid_id,236736.0,5.000000,2.581994,1.000000,3.000000,5.000000,7.000000,9.000000
latitude,236736.0,37.250000,0.204125,37.000000,37.000000,37.250000,37.500000,37.500000
longitude,236736.0,129.000000,0.204125,128.750000,128.750000,129.000000,129.250000,129.250000
heightAboveGround_10_10u,236736.0,1.237709,2.312744,-17.183594,-0.380477,1.246916,2.567676,17.827602
heightAboveGround_10_10v,236736.0,0.177562,1.732229,-24.243887,-0.698129,0.315629,1.064908,15.464080
heightAboveGround_80_u,236736.0,1.928002,3.430303,-21.813850,-0.428749,1.710230,3.616421,25.659668
heightAboveGround_80_v,236736.0,0.222579,2.362594,-30.727304,-0.932746,0.354571,1.365780,19.623650
heightAboveGround_100_100u,236736.0,2.025617,3.597086,-22.420181,-0.436756,1.761622,3.768380,27.149450
heightAboveGround_100_100v,236736.0,0.228532,2.449771,-31.509785,-0.959666,0.356141,1.402044,20.112705
heightAboveGround_2_2t,236736.0,283.562452,10.391815,251.101650,275.275425,284.393560,292.304062,308.400000


[IQR 기준 이상치 후보 개수 (상위 10개 컬럼)]


surface_0_prate                   49897
surface_0_tp                      44519
planetaryBoundaryLayer_0_VRATE    25698
planetaryBoundaryLayer_0_v        25320
planetaryBoundaryLayer_0_u        23938
surface_0_gust                    18044
heightAboveGround_100_100v        14522
heightAboveGround_80_v            14079
isobaricInhPa_850_v               12992
heightAboveGround_10_10v          11461
dtype: int64

===== gfs_train 격자/시각 정합성 =====
grid_id 개수: 9 (기대값 9)
grid_id 개수가 기대값과 다른 forecast_kst_dtm 시각 수: 0
(forecast_kst_dtm, grid_id) 조합 중복 개수: 0
격자별 좌표가 고정되지 않은 grid_id 개수: 0
기간 내 누락된 forecast_kst_dtm 시각 수: 0 / 전체 26304
[data_available_kst_dtm 블록 크기 분포 (정상이면 24시간 단위)]


forecast_kst_dtm
24    9864
Name: count, dtype: int64

## scada_vestas_train 데이터 이슈파악

In [9]:
summarize_basic(scada_vestas_train, "scada_vestas_train")
check_scada(
    scada_vestas_train, "scada_vestas_train",
    prefix="vestas", turbine_ids=[f"wtg{i:02d}" for i in range(1, 13)],
)

===== scada_vestas_train 기본 EDA =====
데이터 크기: (157819, 37)


,kst_dtm,vestas_wtg01_power_kw10m,vestas_wtg02_power_kw10m,vestas_wtg03_power_kw10m,vestas_wtg04_power_kw10m,vestas_wtg05_power_kw10m,vestas_wtg06_power_kw10m,vestas_wtg07_power_kw10m,vestas_wtg08_power_kw10m,vestas_wtg09_power_kw10m,...,vestas_wtg03_wd,vestas_wtg04_wd,vestas_wtg05_wd,vestas_wtg06_wd,vestas_wtg07_wd,vestas_wtg08_wd,vestas_wtg09_wd,vestas_wtg10_wd,vestas_wtg11_wd,vestas_wtg12_wd
0,2022-01-01 01:00:00,325,282,383,470,423,122,200,206,195,...,281.2,308.0,275.6,285.9,270.7,278.1,292.6,247.2,291.0,95.2
1,2022-01-01 01:10:00,338,312,367,420,370,87,153,215,234,...,283.8,307.9,273.8,290.5,268.7,283.1,301.7,253.2,295.3,96.2
2,2022-01-01 01:20:00,333,325,403,476,450,177,210,204,247,...,283.5,307.3,270.8,282.4,265.2,278.8,301.0,252.3,294.7,93.4
3,2022-01-01 01:30:00,356,340,392,453,449,177,170,232,239,...,286.2,315.7,277.4,286.9,273.0,281.8,302.5,252.7,296.2,95.8
4,2022-01-01 01:40:00,417,409,438,487,420,185,175,206,208,...,285.7,318.9,276.5,289.0,273.8,282.4,305.4,252.7,295.7,92.6


[dtypes]


kst_dtm                     datetime64[us]
vestas_wtg01_power_kw10m             int64
vestas_wtg02_power_kw10m             int64
vestas_wtg03_power_kw10m             int64
vestas_wtg04_power_kw10m             int64
vestas_wtg05_power_kw10m             int64
vestas_wtg06_power_kw10m             int64
vestas_wtg07_power_kw10m             int64
vestas_wtg08_power_kw10m             int64
vestas_wtg09_power_kw10m             int64
vestas_wtg10_power_kw10m             int64
vestas_wtg11_power_kw10m             int64
vestas_wtg12_power_kw10m             int64
vestas_wtg01_ws                    float64
vestas_wtg02_ws                    float64
vestas_wtg03_ws                    float64
vestas_wtg04_ws                    float64
vestas_wtg05_ws                    float64
vestas_wtg06_ws                    float64
vestas_wtg07_ws                    float64
vestas_wtg08_ws                    float64
vestas_wtg09_ws                    float64
vestas_wtg10_ws                    float64
vestas_wtg1

[결측치 개수 / 비율(%) - 결측 존재 컬럼만]


,count,pct


[중복 행 개수] 0
[숫자형 컬럼 스케일 확인]


,count,mean,std,min,25%,50%,75%,max
vestas_wtg01_power_kw10m,157819.0,195.472066,712805.870174,-42512957.0,5.0000,102.000,371.0000,4.251296e+07
vestas_wtg02_power_kw10m,157819.0,201.289148,691914.947412,-41959813.0,9.0000,107.000,389.0000,4.195981e+07
vestas_wtg03_power_kw10m,157819.0,217.457809,786801.148854,-46559445.0,12.0000,126.000,431.0000,4.655944e+07
vestas_wtg04_power_kw10m,157819.0,186.633777,480318.380515,-40642699.0,2.0000,97.000,345.0000,4.064270e+07
vestas_wtg05_power_kw10m,157819.0,153.188786,456684.663624,-33717814.0,0.0000,58.000,268.0000,3.371781e+07
vestas_wtg06_power_kw10m,157819.0,158.188127,478422.426755,-34184409.0,3.0000,72.000,274.0000,3.418441e+07
vestas_wtg07_power_kw10m,157819.0,138.377610,418071.325751,-29770834.0,0.0000,52.000,233.0000,2.977083e+07
vestas_wtg08_power_kw10m,157819.0,185.373447,802538.179135,-42301001.0,5.0000,95.000,350.0000,4.230100e+07
vestas_wtg09_power_kw10m,157819.0,199.256205,580428.740812,-43556999.0,5.0000,102.000,385.0000,4.355699e+07
vestas_wtg10_power_kw10m,157819.0,239.206572,842549.632512,-51376454.0,14.0000,149.000,499.0000,5.137645e+07


[IQR 기준 이상치 후보 개수 (상위 10개 컬럼)]


vestas_wtg07_power_kw10m    2769
vestas_wtg06_ws             2026
vestas_wtg09_ws             1802
vestas_wtg04_ws             1500
vestas_wtg07_ws             1413
vestas_wtg03_ws             1379
vestas_wtg02_ws             1288
vestas_wtg08_ws             1244
vestas_wtg12_ws             1155
vestas_wtg01_ws              950
dtype: int64

===== scada_vestas_train SCADA 이슈파악 =====
타임스탬프 중복 개수: 0
기간 내 누락된 타임스탬프 수: 0 / 전체 157819


,결측치,음수 power,동일값 최대 연속 길이,풍속0인데 power>0 건수,풍향 범위
turbine,,,,,
wtg01,0,24784,494,0,0.0 ~ 359.0
wtg02,0,23751,494,0,0.0 ~ 359.0
wtg03,0,21506,494,0,0.0 ~ 359.0
wtg04,0,26723,2776,0,0.0 ~ 359.0
wtg05,0,27255,7470,0,0.0 ~ 359.0
wtg06,0,28489,572,0,0.0 ~ 359.0
wtg07,0,37209,494,0,0.0 ~ 359.0
wtg08,0,28198,494,1,0.0 ~ 359.0
wtg09,0,27372,494,0,0.0 ~ 359.0


## scada_unison_train 데이터 이슈파악

In [10]:
summarize_basic(scada_unison_train, "scada_unison_train")
check_scada(
    scada_unison_train, "scada_unison_train",
    prefix="unison", turbine_ids=[f"wtg{i:02d}" for i in range(1, 6)],
)

===== scada_unison_train 기본 EDA =====
데이터 크기: (105264, 16)


,kst_dtm,unison_wtg01_power_kw10m,unison_wtg02_power_kw10m,unison_wtg03_power_kw10m,unison_wtg04_power_kw10m,unison_wtg05_power_kw10m,unison_wtg01_ws,unison_wtg02_ws,unison_wtg03_ws,unison_wtg04_ws,unison_wtg05_ws,unison_wtg01_wd,unison_wtg02_wd,unison_wtg03_wd,unison_wtg04_wd,unison_wtg05_wd
0,2023-01-01 00:10:00,706.0,702.0,0.0,705.0,707.0,13.73,15.03,16.42,13.97,14.57,-79.005327,-69.946609,-124.739623,-90.938062,-130.941880
1,2023-01-01 00:20:00,706.0,702.0,0.0,684.0,708.0,13.61,14.76,16.51,14.96,14.34,-77.479172,-68.192593,-122.671765,-87.976756,-131.394959
2,2023-01-01 00:30:00,708.0,702.0,0.0,408.0,706.0,14.59,14.70,16.41,12.68,13.55,-75.624282,-67.050591,-122.046216,-87.009944,-132.164142
3,2023-01-01 00:40:00,699.0,384.0,0.0,702.0,706.0,14.62,15.64,15.28,12.46,14.37,-76.608166,-66.836232,-122.675858,-88.533341,-132.392295
4,2023-01-01 00:50:00,610.0,471.0,0.0,698.0,706.0,14.20,15.36,14.79,14.16,14.95,-76.231935,-65.485185,-123.190201,-88.352211,-132.589051


[dtypes]


kst_dtm                     datetime64[us]
unison_wtg01_power_kw10m           float64
unison_wtg02_power_kw10m           float64
unison_wtg03_power_kw10m           float64
unison_wtg04_power_kw10m           float64
unison_wtg05_power_kw10m           float64
unison_wtg01_ws                    float64
unison_wtg02_ws                    float64
unison_wtg03_ws                    float64
unison_wtg04_ws                    float64
unison_wtg05_ws                    float64
unison_wtg01_wd                    float64
unison_wtg02_wd                    float64
unison_wtg03_wd                    float64
unison_wtg04_wd                    float64
unison_wtg05_wd                    float64
dtype: object

[결측치 개수 / 비율(%) - 결측 존재 컬럼만]


,count,pct
unison_wtg01_power_kw10m,208,0.20
unison_wtg02_power_kw10m,1362,1.29
unison_wtg03_power_kw10m,352,0.33
unison_wtg04_power_kw10m,150,0.14
unison_wtg05_power_kw10m,142,0.13
unison_wtg01_ws,1466,1.39
unison_wtg02_ws,1424,1.35
unison_wtg03_ws,447,0.42
unison_wtg04_ws,380,0.36
unison_wtg05_ws,1379,1.31


[중복 행 개수] 0
[숫자형 컬럼 스케일 확인]


,count,mean,std,min,25%,50%,75%,max
unison_wtg01_power_kw10m,105056.0,162.753741,213.965741,0.000000,0.000000,67.000000,289.000000,800.000000
unison_wtg02_power_kw10m,103902.0,145.411022,209.503968,0.000000,0.000000,34.500000,200.000000,1000.000000
unison_wtg03_power_kw10m,104912.0,201.070268,241.747608,0.000000,0.000000,100.000000,395.000000,900.000000
unison_wtg04_power_kw10m,105114.0,196.390233,239.218988,0.000000,0.000000,100.000000,362.750000,800.000000
unison_wtg05_power_kw10m,105122.0,221.543216,262.719079,0.000000,0.000000,100.000000,416.000000,800.000000
unison_wtg01_ws,103798.0,5.509658,3.208078,0.000000,2.810000,4.890000,7.620000,22.160000
unison_wtg02_ws,103840.0,5.183158,3.599933,0.150000,2.500000,4.100000,6.870000,26.520000
unison_wtg03_ws,104817.0,5.996033,3.896123,0.110000,2.730000,5.070000,8.450000,25.040000
unison_wtg04_ws,104884.0,6.049325,3.927545,0.000000,2.770000,5.190000,8.450000,27.980000
unison_wtg05_ws,103885.0,6.428533,4.172445,0.000000,2.910000,5.440000,9.150000,32.050000


[IQR 기준 이상치 후보 개수 (상위 10개 컬럼)]


unison_wtg02_power_kw10m    10682
unison_wtg02_ws              4065
unison_wtg04_ws              1216
unison_wtg03_ws               834
unison_wtg05_ws               772
unison_wtg01_ws               479
unison_wtg01_power_kw10m       19
unison_wtg04_power_kw10m        0
unison_wtg05_power_kw10m        0
unison_wtg03_power_kw10m        0
dtype: int64

===== scada_unison_train SCADA 이슈파악 =====
타임스탬프 중복 개수: 0
기간 내 누락된 타임스탬프 수: 0 / 전체 105264


,결측치,음수 power,동일값 최대 연속 길이,풍속0인데 power>0 건수,풍향 범위
turbine,,,,,
wtg01,208,0,4275,0,-180.0 ~ 180.0
wtg02,1362,0,523,0,-180.0 ~ 180.0
wtg03,352,0,910,0,-180.0 ~ 180.0
wtg04,150,0,866,0,-180.0 ~ 180.0
wtg05,142,0,1726,0,-180.0 ~ 180.0


## info.xlsx 확인

In [11]:
print("데이터 크기:", info_df.shape)
display(info_df)
print("[결측치 개수]")
display(info_df.isnull().sum())
print("[KPX그룹별 터빈 수 / 그룹설비용량 확인]")
display(info_df.groupby("KPX그룹").agg(터빈수=("호기", "count"), 그룹설비용량=("그룹설비용량(MW)", "unique")))

데이터 크기: (17, 11)


,단계,명칭,제작사,모델명,호기,좌표(Google),KPX그룹,Hub Height(m),Rotor Diameter(m),설비용량(MW),그룹설비용량(MW)
0,1,태백가덕산,VESTAS,V126,1,"37°16'55.61""N 128°57'02.10""E",1.0,117,126,3.6,21.6
1,1,태백가덕산,VESTAS,V126,2,"37°17'04.05""N 128°56'58.35""E",1.0,117,126,3.6,21.6
2,1,태백가덕산,VESTAS,V126,3,"37°17'11.49""N 128°56'58.99""E",1.0,117,126,3.6,21.6
3,1,태백가덕산,VESTAS,V126,4,"37°17'23.11""N 128°57'03.68""E",1.0,117,126,3.6,21.6
4,1,태백가덕산,VESTAS,V126,5,"37°17'28.20""N 128°57'15.58""E",1.0,117,126,3.6,21.6
5,1,태백가덕산,VESTAS,V126,6,"37°17'19.48""N 128°57'24.96""E",1.0,117,126,3.6,21.6
6,1,태백가덕산,VESTAS,V126,7,"37°17'16.20""N 128°57'34.67""E",2.0,117,126,3.6,21.6
7,1,태백가덕산,VESTAS,V126,8,"37°17'11.29""N 128°57'47.24""E",2.0,117,126,3.6,21.6
8,1,태백가덕산,VESTAS,V126,9,"37°17'00.97""N 128°57'57.44""E",2.0,117,126,3.6,21.6
9,1,태백가덕산,VESTAS,V126,10,"37°16'52.77""N 128°58'04.18""E",2.0,117,126,3.6,21.6


[결측치 개수]


단계                   0
명칭                   0
제작사                  0
모델명                  0
호기                   0
좌표(Google)           0
KPX그룹                0
Hub Height(m)        0
Rotor Diameter(m)    0
설비용량(MW)             0
그룹설비용량(MW)           0
dtype: int64

[KPX그룹별 터빈 수 / 그룹설비용량 확인]


,터빈수,그룹설비용량
KPX그룹,,
1.0,6,[21.6]
2.0,6,[21.6]
3.0,5,[21.0]


## train / test 정합성 및 설비용량 초과 확인

In [12]:
print("ldaps_train / ldaps_test 컬럼 차이:", set(ldaps_train.columns) ^ set(ldaps_test.columns))
print("gfs_train / gfs_test 컬럼 차이:", set(gfs_train.columns) ^ set(gfs_test.columns))

check_forecast_grid(ldaps_test, "ldaps_test", expected_grid_count=16)
check_forecast_grid(gfs_test, "gfs_test", expected_grid_count=9)

print("===== train_labels 발전량 vs 설비용량 초과/음수 확인 =====")
for col in TARGET_COLS:
    over = (train_labels[col] > CAPACITY_KWH[col]).sum()
    neg = (train_labels[col] < 0).sum()
    print(f"{col}: 설비용량({CAPACITY_KWH[col]} kWh) 초과 {over}건, 음수 {neg}건")

ldaps_train / ldaps_test 컬럼 차이: set()
gfs_train / gfs_test 컬럼 차이: set()
===== ldaps_test 격자/시각 정합성 =====
grid_id 개수: 16 (기대값 16)
grid_id 개수가 기대값과 다른 forecast_kst_dtm 시각 수: 0
(forecast_kst_dtm, grid_id) 조합 중복 개수: 0
격자별 좌표가 고정되지 않은 grid_id 개수: 0
기간 내 누락된 forecast_kst_dtm 시각 수: 0 / 전체 8760
[data_available_kst_dtm 블록 크기 분포 (정상이면 24시간 단위)]


forecast_kst_dtm
24    5840
Name: count, dtype: int64

===== gfs_test 격자/시각 정합성 =====
grid_id 개수: 9 (기대값 9)
grid_id 개수가 기대값과 다른 forecast_kst_dtm 시각 수: 0
(forecast_kst_dtm, grid_id) 조합 중복 개수: 0
격자별 좌표가 고정되지 않은 grid_id 개수: 0
기간 내 누락된 forecast_kst_dtm 시각 수: 0 / 전체 8760
[data_available_kst_dtm 블록 크기 분포 (정상이면 24시간 단위)]


forecast_kst_dtm
24    3285
Name: count, dtype: int64

===== train_labels 발전량 vs 설비용량 초과/음수 확인 =====
kpx_group_1: 설비용량(21600 kWh) 초과 0건, 음수 0건
kpx_group_2: 설비용량(21600 kWh) 초과 0건, 음수 0건
kpx_group_3: 설비용량(21000 kWh) 초과 38건, 음수 0건


## 이슈파악 요약

- **LDAPS/GFS 예보**: 결측치 없음, grid_id 개수(16/9) 및 좌표 고정 정상, `forecast_kst_dtm` 누락 없음, `data_available_kst_dtm` 24시간 블록 규칙 정상, train/test 컬럼 스키마 일치.
- **train_labels**: `kpx_group_3`이 그룹 설비용량(21,000 kWh)을 초과하는 값이 38건 존재 → 후처리(스케일 보정)나 원인 분석이 필요한 이상치 후보.
- **scada_vestas_train**: `power_kw10m` 컬럼에 정격출력(3.6 MW)로는 설명되지 않는 초대형 이상치(최대 약 ±5×10⁷)가 다수(1~2만 건/터빈) 존재. 정상 범위를 크게 벗어난 값이라 센서 오류/에러 코드일 가능성이 높고, 그대로 사용하면 학습에 치명적 → 클렌징(예: 정격출력 기준 클리핑 또는 결측 처리) 필수.
- **scada_unison_train**: 컬럼별 결측치 0.1~1.4% 존재, `wtg02_power`에 이상치 다수, 풍향(`wd`) 범위가 VESTAS(0~359)와 달리 -180~180로 스케일이 달라 통합 시 변환 필요.
- **info.xlsx**: KPX그룹별 터빈 수(6/6/5)·그룹설비용량(21.6/21.6/21.0 MW)이 `data_description.md` 및 `TARGET_COLS` 설비용량과 일치함을 확인.
